test/gen_types.zig

```diff
@@ -0,0 +1,260 @@
+};
+
+// Check duplicate types
+comptime {
+    @setEvalBranchQuota(types.len * types.len * 7);
+    inline for (types) |t1, i| {
+        inline for (types) |t2, j| {
+            if (i != j) {
+                if (std.mem.eql(u8, t1[0], t2[0])) {
+                    @compileError("Duplicate types: " ++ t1[0]);
+                } else if (std.mem.eql(u8, t1[1], t2[1])) {
+                    @compileError("Duplicate enum literal: " ++ t1[1]);
+                }
+            }
+        }
+    }
+}
+
+// Create the imports
+fn genImports() []const u8 {
+    @setEvalBranchQuota(types.len * types.len * 7);
+    comptime var str: []const u8 = "";
+    comptime var seen_imports: []const u8 = &[_]u8{};
+    comptime var seen_types: []const u8 = &[_]u8{};
+
+    inline for (types) |t| {
+        const has_import = !std.mem.eql(u8, t[2], "");
+        const seen = if (std.mem.indexOf(u8, seen_imports, t[2])) |_| true else false;
+        if (has_import and !seen) {
+            str = str ++ "const " ++ t[2] ++ " = @import(\"" ++ t[2] ++ ".zig\");\n";
+            seen_imports = seen_imports ++ t[2];
+        }
+    }
+
+    inline for (types) |t| {
+        const has_import = !std.mem.eql(u8, t[2], "");
+        const has_base = !std.mem.eql(u8, t[3], "");
+        const has_type = !std.mem.eql(u8, t[4], "");
+        const seen = if (std.mem.indexOf(u8, seen_types, t[4])) |_| true else false;
+        if (!seen and has_type and (has_import or has_base)) {
+            str = str ++ "const " ++ t[4] ++ " = ";
+            if (has_import) {
+                str = str ++ t[2] ++ ".";
+            }
+            if (has_base) {
+                str = str ++ t[3] ++ ".";
+            }
+            str = str ++ t[4] ++ ";\n";
+            seen_types = seen_types ++ t[4];
+        }
+    }
+    // Remove trailing new line
+    return str;
+}
+
+// Create the DataElementType
+fn genDataElementType() []const u8 {
+    comptime var str: []const u8 = "const DataElementType = enum {\n";
+    inline for (types) |t| {
+        str = str ++ "    " ++ t[1] ++ ",\n";
+    }
+    return str ++ "};\n";
+}
+
+// Create the DataElement
+fn genDataElement() []const u8 {
+    comptime var str: []const u8 = "const DataElement = union(DataElementType) {\n";
+    inline for (types) |t| {
+        str = str ++ "    " ++ t[1] ++ ": " ++ t[0] ++ ",\n";
+    }
+    return str ++ "};\n";
+}
+
+// All the function generation parts are the same apart from 3 things
+fn genGenericFunc(comptime intermediate: []const u8, comptime trail: []const u8, comptime end: []const u8) []const u8 {
+    comptime var str: []const u8 = "";
+    inline for (types) |t, i| {
+        if (i == 0) {
+            // Want nice formatting
+            str = str ++ "    " ++ t[0] ++ intermediate ++ t[1] ++ trail;
+        } else {
+            str = str ++ "                " ++ t[0] ++ intermediate ++ t[1] ++ trail;
+        }
```

Comment:
We could probably remove the duplicate code by assigning the big ol' block of spaces to a const and using that.